job-runner/runner/runner.go

```diff
@@ -165,27 +176,32 @@ TRY_LOOP:
 			break TRY_LOOP
 		}
 
+		if jobRet.State == proto.STATE_STOPPED {
+			tryLogger.Info("job stopped")
+			break TRY_LOOP
+		}
+
 		// //////////////////////////////////////////////////////////////////
 		// Job failed, wait and retry?
 		// //////////////////////////////////////////////////////////////////
-
-		// Log the failure
-		tryLogger.Errorf("job failed because state != %s: state = %s",
-			proto.StateName[proto.STATE_COMPLETE], proto.StateName[jl.State])
+		tryLogger.Errorf("job failed: state %s (%d)", proto.StateName[jl.State], jl.State)
 
 		// If last try, break retry loop, don't wait
 		if tryNo == r.maxTries {
 			break TRY_LOOP
 		}
+		tryNo++ // see next comment...
 
-		// Wait between retries
+		// Wait between retries. Can be stopped while waiting which is why we
+		// need to increment tryNo first. At this point, we're effectively on
+		// the next try. E.g. try 1 fails, we're waiting for try 2, then we're
+		// stopped: we want try 2 state = stopped so that on resume try 2 is re-ran.
 		select {
 		case <-time.After(r.retryWait):
 		case <-r.stopChan:
-			// runner has been stopped
+			tryLogger.Infof("job stopped while waiting to run try %d", tryNo)
 			break TRY_LOOP
```

Comment:
Hmmmmm, because we return with state Failed here (from the last run), I'm pretty sure the reaper will prepare a sequence retry, even if there are some job-level retries left. Setting final state = Stopped should fix this.

Reply:
Yes, thanks. Supposed to be a `finalState = proto.STATE_STOPPED` there.